### ISSUE TITLE
Different results for single sample vs. batched

### ISSUE THREAD

**Issue by rightaditya** [pytorch/pytorch#124025]
### 🐛 Describe the bug

I've been running into an issue where our code yields different results depending on whether a sample is submitted on its own or as part of a batch, and in our case it's important that the same input always yields the same predicted outputs. I'm including some examples below. In case the particular tensors from my examples are useful, they're available in [this tarball](https://github.com/pytorch/pytorch/files/14972492/xw.tar.gz).

The examples below were run on my laptop, except for the CUDA one (did that on an A10G), but I've replicated this on an Intel Sapphire Rapids CPU (Xeon Platinum 8488C) as well. The differences are small here, but they compound over successive operations, and can thus result in different output predictions, especially (but not only) when the top predicted classes are close in their scores. Some minor differences are understandable, but we've found over 10% of sentences have a different tag in them somewhere based solely on whether they were submitted one at a time vs. batched. And I would've expected such differences to pass `torch.allclose`, but as shown below, they don't for CUDA and only barely do for CPU.

Is this magnitude of difference within expectations? If so, are there any recommendations for working around them?

A minimal example:
```python
>>> import torch
>>> x = torch.randn(1, 32)
>>> w = torch.randn(32, 1)
>>> (x @ w)[0, 0].item()
-0.046955958008766174
>>> (x.expand(32, -1) @ w)[0, 0].item()
-0.04695606231689453
```

In this case, the difference passes `torch.allclose` on CPU, but just barely:
```python
>>> torch.allclose(x @ w, x.expand(32, -1) @ w)
True
>>> torch.allclose(x @ w, x.expand(32, -1) @ w, rtol=1e-6)
False
```
And it doesn't pass on CUDA:
```python
>>> torch.allclose(x.to('cuda') @ w.to('cuda'), x.to('cuda').expand(32, -1) @ w.to('cuda'))
False
```

Expanding `w` fixes things (or at least, gives the same result as non-batched):
```python
>>> (x.expand(32, -1) @ w.expand(-1, 32))[0, 0].item()
-0.046955958008766174
```

But then if `w` is transposed, the difference comes back:
```python
>>> (x.expand(32, -1) @ w.expand(-1, 32).T)[0, 0].item()
-4.249419689178467
>>> (x @ w.expand(-1, 32).T)[0, 0].item()
-4.24941873550415
```

This also affects `F.linear`:
```python
>>> import torch.nn.functional as F
>>> F.linear(x, w.expand(-1, 32))[0, 0].item()
-4.24941873550415
>>> F.linear(x.expand(32, -1), w.expand(-1, 32))[0, 0].item()
-4.249419689178467
```

Using fp64 seems to help, though for us would be too slow to be practical:
```python
>>> (x.double() @ w.double())[0, 0].item()
-0.046956273061034004
>>> (x.double().expand(32, -1) @ w.double())[0, 0].item()
-0.04695627306103356
```

fp16 seems to have no issues on these particular inputs:
```python
>>> (x.half() @ w.half())[0, 0].item()
-0.046966552734375
>>> (x.half().expand(32, -1) @ w.half())[0, 0].item()
-0.046966552734375
>>> (x.half().expand(32, -1) @ w.half().expand(-1, 32))[0, 0].item()
-0.046966552734375
>>> (x.half().expand(32, -1) @ w.half().expand(-1, 32).T)[0, 0].item()
-4.25
>>> (x.half() @ w.half().expand(-1, 32).T)[0, 0].item()
-4.25
```

And bf16 is likewise consistent, though egregiously off in this case:
```python
>>> (x.bfloat16() @ w.bfloat16())[0, 0].item()
0.001495361328125
>>> (x.bfloat16().expand(32, -1) @ w.bfloat16())[0, 0].item()
0.001495361328125
>>> (x.bfloat16().expand(32, -1) @ w.bfloat16().expand(-1, 32))[0, 0].item()

0.001495361328125
>>> (x.bfloat16().expand(32, -1) @ w.bfloat16().expand(-1, 32).T)[0, 0].item()
-4.25
>>> (x.bfloat16() @ w.bfloat16().expand(-1, 32).T)[0, 0].item()
-4.25
```

**NB:** I haven't tested these more extensively, so it's possible that fp16 and/or bf16 do exhibit this issue for some other inputs.

### Versions

```
PyTorch version: 2.2.2+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Manjaro Linux (x86_64)
GCC version: (GCC) 13.2.1 20230801
Clang version: 16.0.6
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.11.8 (main, Mar 21 2024, 21:38:22) [GCC 13.2.1 20230801] (64-bit runtime)
Python platform: Linux-6.6.19-1-MANJARO-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      39 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             8
On-line CPU(s) list:                0-7
Vendor ID:                          GenuineIntel
Model name:                         11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
CPU family:                         6
Model:                              140
Thread(s) per core:                 2
Core(s) per socket:                 4
Socket(s):                          1
Stepping:                           1
CPU(s) scaling MHz:                 69%
CPU max MHz:                        4200.0000
CPU min MHz:                        400.0000
BogoMIPS:                           4840.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l2 cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid movdiri movdir64b fsrm avx512_vp2intersect md_clear ibt flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          192 KiB (4 instances)
L1i cache:                          128 KiB (4 instances)
L2 cache:                           5 MiB (4 instances)
L3 cache:                           8 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-7
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] torch==2.2.2+cpu
[conda] Could not collect
```

**Comment by zou3519** [comment]
This is expected; see https://pytorch.org/docs/stable/notes/numerical_accuracy.html#batched-computations-or-slice-computations

**Comment by rightaditya** [comment]
@zou3519 Thanks, but I'm not expecting bitwise-identical results, which is what that snippet seems to be focusing on. Is there some maximum difference that's targeted for these kinds of differences? Surely it's not allowed to be arbitrarily large. In particular, I wouldn't expect `torch.allclose` to fail for any of these, especially for higher-precision cases.

If the default settings of `torch.allclose` are too small an acceptable upper bound, it really seems like that section should be reflect this. For example, it says "`(A@B)[0]` ... is not guaranteed to be bitwise identical to `A[0]@B[0]`". That makes it sound like the difference should be minuscule. Perhaps it should be expressed as "`(A@B)[0]` ... is not guaranteed to be close to `A[0]@B[0]`"?

**Comment by zou3519** [comment]
> Is there some maximum difference that's targeted for these kinds of differences? 

The doc says "In particular, note that floating point provides limited accuracy (about 7 decimal digits for single precision floating point numbers, about 16 decimal digits for double precision floating point numbers) "

**Comment by rightaditya** [comment]
@zou3519 That's not what I'm asking, unless you're suggesting that the maximum observed difference between single-sample and batched results should be expected to be accurate within "about 7 decimal digits for single precision floating point numbers." I don't think that's what you're suggesting, nor what the doc is talking about, but if it is—the examples I've found clearly show differences outside this and should be classed as a bug.

What I'm asking is: how are the results of batched ops validated against single-sample ops? If someone comes up with a new implementation for a batched operation, how would it be tested to make sure it's adequately close to the single-sample version? What would be the maximum allowable deviation?

**Comment by Zoher15** [comment]
Why was this issue closed? It seems like a huge issue. How are you dealing with it @rightaditya ?

**Comment by Zoher15** [comment]
@zou3519 maybe this needs to be reopened?

**Comment by zou3519** [comment]
I'm happy to reopen this issue and take another look.  Let me try to make a pass through the unanswered questions through the thread. If you folks have specific repros (other than just the matmul repros in the original issue -- those seemed reasonable to me), please post them here.

> What I'm asking is: how are the results of batched ops validated against single-sample ops? If someone comes up with a new implementation for a batched operation, how would it be tested to make sure it's adequately close to the single-sample version? What would be the maximum allowable deviation?

We test our operators against reference implementations during testing. Most of them pass in the torch.allclose allowed precision. Sometimes it's an art and we allow thinner tolerances (1e-5, 1e-4) for fp32.

Matmuls, especially on CUDA, are interesting. The cublas libraries we call into are very non-deterministic. If you want them to be deterministic, you'll have to play with their settings. The first thing to try is https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html
